typescript/agentkit/src/wallet-providers/smartWalletProvider.ts

```diff
@@ -0,0 +1,372 @@
+import {
+  CHAIN_ID_TO_NETWORK_ID,
+  Coinbase,
+  createSmartWallet,
+  NetworkScopedSmartWallet,
+  SendUserOperationOptions,
+  Signer,
+  SupportedChainId,
+  toSmartWallet,
+  waitForUserOperation,
+} from "@coinbase/coinbase-sdk";
+import {
+  Abi,
+  ContractFunctionArgs,
+  ContractFunctionName,
+  createPublicClient,
+  Hex,
+  http,
+  Prettify,
+  ReadContractParameters,
+  ReadContractReturnType,
+  TransactionRequest,
+  PublicClient as ViemPublicClient,
+} from "viem";
+import { Network, NETWORK_ID_TO_CHAIN_ID, NETWORK_ID_TO_VIEM_CHAIN } from "../network";
+import { EvmWalletProvider } from "./evmWalletProvider";
+
+export interface ConfigureSmartWalletWithSignerOptions {
+  cdpApiKeyName?: string;
+  cdpApiKeyPrivateKey?: string;
+  networkId?: string;
+  chainId?: string;
+  smartWalletAddress?: Hex;
+  signer: Signer;
+}
+
+interface SmartWalletProviderConfig {
+  smartWallet: NetworkScopedSmartWallet;
+  network: Required<Network>;
+  chainId: string;
+}
+
+/**
+ *
+ */
+export class SmartWalletProvider extends EvmWalletProvider {
+  #smartWallet: NetworkScopedSmartWallet;
+  #network: Required<Network>;
+  #publicClient: ViemPublicClient;
+
+  /**
+   * Constructs a new CdpWalletProvider.
+   *
+   * @param config - The configuration options for the CdpWalletProvider.
+   */
+  private constructor(config: SmartWalletProviderConfig) {
+    super();
+
+    this.#network = config.network;
+    this.#smartWallet = config.smartWallet;
+    this.#publicClient = createPublicClient({
+      chain: NETWORK_ID_TO_VIEM_CHAIN[config.network.networkId],
+      transport: http(),
+    });
+  }
+
+  /**
+   * Configures and returns a `SmartWalletProvider` instance using the provided configuration options.
+   * This method initializes a smart wallet based on the given network and credentials.
+   *
+   * @param {ConfigureSmartWalletWithSignerOptions} config
+   *   - Configuration parameters for setting up the smart wallet.
+   *
+   * @returns {Promise<SmartWalletProvider>}
+   *   - A promise that resolves to an instance of `SmartWalletProvider` configured with the provided settings.
+   *
+   * @throws {Error}
+   *   - If an invalid combination of `networkId` and `chainId` is provided.
+   *   - If the `chainId` cannot be determined.
+   *   - If the `chainId` is not supported.
+   *   - If `CDP_API_KEY_NAME` or `CDP_API_KEY_PRIVATE_KEY` is missing.
+   *
+   * @example
+   * ```typescript
+   * const smartWalletProvider = await SmartWalletProvider.configureWithWallet({
+   *   networkId: "base-sepolia",
+   *   signer: privateKeyToAccount("0xethprivatekey"),
+   *   cdpApiKeyName: "my-api-key",
+   *   cdpApiKeyPrivateKey: "my-private-key",
+   *   smartWalletAddress: "0x123456...",
+   * });
+   * ```
+   */
+  public static async configureWithSigner(
```

Comment:
Wonder if we should prefer `configureWithWallet` for consistency across other providers